core/src/block_sync.rs

```diff
@@ -240,10 +240,10 @@ pub mod message {
                 }
                 Message::ShareBlocks(ShareBlocks { blocks, .. }) => {
                     use crate::sumeragi::message::{BlockSyncUpdate, Message, MessagePacket};
-                    for block in blocks {
+                    for block in blocks.clone() {
                         block_sync.sumeragi.incoming_message(MessagePacket::new(
                             Vec::new(),
-                            Message::BlockSyncUpdate(BlockSyncUpdate::from(block.clone())),
+                            Message::BlockSyncUpdate(BlockSyncUpdate { block }),
```

Comment:
is this some leftover from the actor system? why do we send one by one block to sumeragi?

Reply:
I tried refactoring it under the assumption that it was a leftover. It ended up costing around 20% in performance. I suspect it's because of the prevalence of `Mutex` es in the `main_loop` which means that blocking for a significant amount of time can pessimism performance greatly. 

With that said, we might want to investigate other synchronisation methods. I'm willing to wager that Atomic pointers with an arena allocator would do very well in terms of performance.

Reply:
good, as long it's been brought to our attention we can look into it in the future

Reply:
Originally sumeragi recieved BlockCommitted messages that originated both from other peers and block_sync. We could change it although I doubt it would help much since performance is dominated by WSV::apply.